### PR TITLE
Download dataset with given version

### PR DESCRIPTION
**Pull request by vijayphoenix**
Fix https://github.com/tensorflow/datasets/issues/1707

### CODE REVIEW

File: tensorflow_datasets/scripts/download_and_prepare.py [hunk 1]
```diff
@@ -156,6 +156,13 @@ def import_modules(modules):
   for module in modules:
     importlib.import_module(module)
 
+def _get_version(builder_name, version = None):
```
Comment (Conchylicultor) [inline tensorflow_datasets/scripts/download_and_prepare.py]: We should not duplicate logic. This code is already present in `tfds.builder` as `tfds.builder('ds:version')` works.
Could you find another way ?

**Comment by vijayphoenix** [comment]
```
(base) ➜  image_classification git:(version) python -m tensorflow_datasets.scripts.download_and_prepare --datasets=scientific_papers:1.1.0
I0325 03:28:55.778770 139847452337984 download_and_prepare.py:188] Running download_and_prepare for datasets:
scientific_papers:1.1.0
I0325 03:28:55.778964 139847452337984 download_and_prepare.py:189] Version: "None"
I0325 03:28:55.779296 139847452337984 dataset_builder.py:785] No config specified, defaulting to first: scientific_papers/arxiv
I0325 03:28:55.779478 139847452337984 dataset_builder.py:203] Load pre-computed datasetinfo (eg: splits) from bucket.
I0325 03:28:56.979401 139847452337984 dataset_info.py:432] Loading info from GCS for scientific_papers/arxiv/1.1.0
I0325 03:28:57.340367 139847452337984 dataset_builder.py:203] Load pre-computed datasetinfo (eg: splits) from bucket.
I0325 03:28:57.341478 139847452337984 dataset_info.py:432] Loading info from GCS for scientific_papers/arxiv/1.1.0
I0325 03:28:57.699569 139847452337984 download_and_prepare.py:131] download_and_prepare for dataset scientific_papers/arxiv/1.1.0...
I0325 03:28:57.700240 139847452337984 dataset_builder.py:311] Generating dataset scientific_papers (/home/vijay/tensorflow_datasets/scientific_papers/arxiv/1.1.0)
Downloading and preparing dataset scientific_papers/arxiv/1.1.0 (download: 4.20 GiB, generated: Unknown size, total: 4.20 GiB) to /home/vijay/tensorflow_datasets/scientific_papers/arxiv/1.1.0...
Dl Completed...: 0 url [00:00, ? url/s]          I0325 03:28:58.616502 139847452337984 download_manager.py:257] Downloading https://drive.google.com/uc?id=1b3rmCSIoh6VhD4HKWjI4HOW-cSwcwbeC&export=download into /home/vijay/tensorflow_datasets/downloads/ucid_1b3rmCSIoh6VhD4H-cSwcwbeC_export_downloadwN6uevfZyH8l3632IfcSb3CNfcrG01PHVkiDCEoAAHY.tmp.8f8fa2538998489bb0a07489a67f60a0...ile/s]
Dl Completed...:   0%|                           I0325 03:28:58.620622 139847452337984 download_manager.py:257] Downloading https://drive.google.com/uc?id=1lvsqvsFi3W-pE1SqNZI0s8NR9rC1tsja&export=download into /home/vijay/tensorflow_datasets/downloads/ucid_1lvsqvsFi3W-pE1SqNZI0s8NR_export_downloadY_jZrsD4nW0oeCgmL5TDLaYprWnpe0-DuXkeCnmmgwQ.tmp.9e8181d9352145c0bac9804dce7b170d...ile/s]
Dl Completed...:   0%|                           /home/vijay/.local/lib/python3.7/site-packages/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made to host 'drive.google.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warningsompleted...: 0 file [00:00, ? file/s]
```

**Comment by vijayphoenix** [comment]
```
(base) ➜  image_classification git:(version) python -m tensorflow_datasets.scripts.download_and_prepare --datasets=scientific_papers:1.1.5
I0325 03:29:50.704036 140249921926976 download_and_prepare.py:188] Running download_and_prepare for datasets:
scientific_papers:1.1.5
I0325 03:29:50.704234 140249921926976 download_and_prepare.py:189] Version: "None"
I0325 03:29:50.704534 140249921926976 dataset_builder.py:785] No config specified, defaulting to first: scientific_papers/arxiv
E0325 03:29:50.704650 140249921926976 registered.py:177] Failed to construct dataset scientific_papers
Traceback (most recent call last):
  File "/home/vijay/miniconda3/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/vijay/miniconda3/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/scripts/download_and_prepare.py", line 228, in <module>
    app.run(main)
  File "/home/vijay/.local/lib/python3.7/site-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/home/vijay/.local/lib/python3.7/site-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/scripts/download_and_prepare.py", line 192, in main
    for name in datasets_to_build
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/scripts/download_and_prepare.py", line 192, in <dictcomp>
    for name in datasets_to_build
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/core/registered.py", line 175, in builder
    return _DATASET_REGISTRY[name](**builder_kwargs)
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/core/api_utils.py", line 53, in disallow_positional_args_dec
    return fn(*args, **kwargs)
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/core/dataset_builder.py", line 196, in __init__
    self._version = self._pick_version(version)
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/core/dataset_builder.py", line 244, in _pick_version
    raise AssertionError(msg)
AssertionError: Dataset scientific_papers cannot be loaded at version 1.1.5, only: 1.1.1, 1.1.0.
(base) ➜  image_classification git:(version)
```

File: tensorflow_datasets/scripts/download_and_prepare.py (hunk 1 above)
Comment (Conchylicultor) [reply tensorflow_datasets/scripts/download_and_prepare.py]: For instance, you could check that passing `tfds.builder(name, data_dir=FLAGS.data_dir, **version_kwargs)` works.
when `version_kwargs = {}` when `FLAGS.version is None`.

**Comment by vijayphoenix** [comment]
New logs
```
(base) ➜  datasets git:(version) python -m tensorflow_datasets.scripts.download_and_prepare --datasets=scientific_papers:1.1.5
I0326 04:21:19.362181 140500197586752 download_and_prepare.py:181] Running download_and_prepare for datasets:
scientific_papers:1.1.5
I0326 04:21:19.362375 140500197586752 download_and_prepare.py:182] Version: "None"
I0326 04:21:19.362670 140500197586752 dataset_builder.py:785] No config specified, defaulting to first: scientific_papers/arxiv
E0326 04:21:19.362782 140500197586752 registered.py:178] Failed to construct dataset scientific_papers
Traceback (most recent call last):
  File "/home/vijay/miniconda3/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/vijay/miniconda3/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/scripts/download_and_prepare.py", line 220, in <module>
    app.run(main)
  File "/home/vijay/.local/lib/python3.7/site-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/home/vijay/.local/lib/python3.7/site-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/scripts/download_and_prepare.py", line 185, in main
    for name in datasets_to_build
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/scripts/download_and_prepare.py", line 185, in <dictcomp>
    for name in datasets_to_build
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/core/registered.py", line 176, in builder
    return _DATASET_REGISTRY[name](**builder_kwargs)
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/core/api_utils.py", line 53, in disallow_positional_args_dec
    return fn(*args, **kwargs)
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/core/dataset_builder.py", line 196, in __init__
    self._version = self._pick_version(version)
  File "/home/vijay/GitHub-Repos/Tensorflow/datasets/tensorflow_datasets/core/dataset_builder.py", line 244, in _pick_version
    raise AssertionError(msg)
AssertionError: Dataset scientific_papers cannot be loaded at version 1.1.5, only: 1.1.1, 1.1.0.
```

**Comment by vijayphoenix** [comment]
On running
```
import tensorflow_datasets as tfds
builder = tfds.builder("scientific_papers:1.1.0",version=None)
builder.download_and_prepare()
```
Logs
```
Downloading and preparing dataset scientific_papers/arxiv/1.1.0 (download: 4.20 GiB, generated: Unknown size, total: 4.20 GiB) to /home/vijay/tensorflow_datasets/scientific_papers/arxiv/1.1.0...
Dl Completed...:   0%|                           /home/vijay/.local/lib/python3.7/site-packages/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made to host 'drive.google.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/home/vijay/.local/lib/python3.7/site-packages/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made to host 'drive.google.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
Extraction completed...: 0 file [00:00, ? file/s]
Dl Size...: 0 MiB [00:00, ? MiB/s]
Dl Completed...:   0%|                                     | 0/2 [00:00<?, ? url/s]
```

**Comment by vijayphoenix** [comment]
But this fix still does not work for 
```
python -m tensorflow_datasets.scripts.download_and_prepare --datasets=scientific_papers:1.1.0
```
I will look into this soon.

File: tensorflow_datasets/core/registered.py [hunk 2]
```diff
@@ -109,7 +109,7 @@ def __init__(self, name, is_abstract=False, in_development=False):
 class RegisteredDataset(abc.ABCMeta):
   """Subclasses will be registered and given a `name` property."""
 
-  def __new__(cls, cls_name, bases, class_dict):
+  def __new__(cls, cls_name, bases, class_dict): # pylint: disable= C0204, W0221
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/registered.py]: Please use the human readable names, rather than symbols `C0204` (e.g. `import-not-on-top`, `variable-redefined`)

File: tensorflow_datasets/core/registered.py [hunk 3]
```diff
@@ -164,7 +164,8 @@ def builder(name, **builder_init_kwargs):
     DatasetNotFoundError: if `name` is unrecognized.
   """
   name, builder_kwargs = _dataset_name_and_kwargs_from_name_str(name)
-  builder_kwargs.update(builder_init_kwargs)
+  builder_kwargs.update( \
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/registered.py]: I may be wrong but I don't think you should not modify `register.py`. Unless generating `tfds.load('scientific_paper:1.1.0')` generate `1.1.1`.

Comment (vijayphoenix) [reply tensorflow_datasets/core/registered.py]: `tfds.builder('scientific_paper:1.1.0', version = None)` generates `1.1.1`

Comment (vijayphoenix) [reply tensorflow_datasets/core/registered.py]: Just read the documentation.
```
**builder_init_kwargs: `dict` of keyword arguments passed to the
   `DatasetBuilder`. These will override keyword arguments passed in `name`,
    if any.
```
So, I fixed this error in the script.

**Comment by vijayphoenix** [comment]
New logs :tada: 
```
(base) ➜  datasets git:(version) python -m tensorflow_datasets.scripts.download_and_prepare --datasets=scientific_papers:1.1.0
I0326 20:18:02.766934 140660102723392 download_and_prepare.py:181] Running download_and_prepare for datasets:
scientific_papers:1.1.0
I0326 20:18:02.767130 140660102723392 download_and_prepare.py:182] Version: "None"
I0326 20:18:02.767464 140660102723392 dataset_builder.py:785] No config specified, defaulting to first: scientific_papers/arxiv
I0326 20:18:02.767690 140660102723392 dataset_builder.py:203] Load pre-computed datasetinfo (eg: splits) from bucket.
I0326 20:18:04.163312 140660102723392 dataset_info.py:432] Loading info from GCS for scientific_papers/arxiv/1.1.0
I0326 20:18:04.561695 140660102723392 dataset_builder.py:203] Load pre-computed datasetinfo (eg: splits) from bucket.
I0326 20:18:04.562705 140660102723392 dataset_info.py:432] Loading info from GCS for scientific_papers/arxiv/1.1.0
I0326 20:18:05.033975 140660102723392 download_and_prepare.py:131] download_and_prepare for dataset scientific_papers/arxiv/1.1.0...
I0326 20:18:05.034722 140660102723392 dataset_builder.py:311] Generating dataset scientific_papers (/home/vijay/tensorflow_datasets/scientific_papers/arxiv/1.1.0)
Downloading and preparing dataset scientific_papers/arxiv/1.1.0 (download: 4.20 GiB, generated: Unknown size, total: 4.20 GiB) to /home/vijay/tensorflow_datasets/scientific_papers/arxiv/1.1.0...
```

**Comment by vijayphoenix** [comment]
https://github.com/tensorflow/datasets/blob/32415bb3bc02f7d6257bb484e9c6d97b159b15e6/tensorflow_datasets/scripts/download_and_prepare.py#L216
It should be
```
          builder_for_config = tfds.builder(
              name,
              data_dir=FLAGS.data_dir,
              config=config,
              **version_kwarg)
```
because `builder.name` is `scientific_papers` 
and `name` is `scientific_papers:1.1.0`

**Comment by vijayphoenix** [comment]
I just realized one more thing, I forgot to update.

**Comment by Conchylicultor** [comment]
@vijayphoenix I'm not sure I understand, could you rephrase what is the issue ?

**Comment by vijayphoenix** [comment]
@Conchylicultor please see the update code and explanation https://github.com/tensorflow/datasets/pull/1734
